### PR TITLE
Fix for 974

### PR DESCRIPTION
**Pull request by PetrDlouhy**
**Problem**

#974

**Solution**

This is concept of a solution of the problem with unnecessary queries for IDs.

I would like to get some feedback on this. I am not sure about which other cases could this affect and whether we are able to fix more cases.

The current solution is mostly bypassing the logic of Widget and leaking it to the Fields. But I am not sure how to make this correctly, because the data in Widgets are not sufficient for this and also when the `Widget.render()` is called the demage is already done and the query is called.

**Acceptance Criteria**

It includes a clear test case scenario documenting the problem.

### CODE REVIEW

**Comment by PetrDlouhy** [comment]
Now I rewrote this to a cleaner solution - instead of changing Widget and Fields logic I tried to change the way the default fields are derived from Model.
I find this solution to be much cleaner and I think I am satisfied with this approach although there is still some tuning needed.

@matthewhegarty Thanks, I rebased against v4.

**Comment by matthewhegarty** [comment]
Yes, in v4 you have to add the decorator to your test:

```
@ignore_widget_deprecation_warning
```

**Comment by matthewhegarty** [comment]
Also make me think that this fix could help with #1209

**Comment by matthewhegarty** [comment]
@PetrDlouhy I tested imports / exports with this fix in place and it seems fine.

One thing I noticed is that we can remove the additional queries using `select_related()`.  This is mentioned in the issue, and most Django users will be familiar with this pattern.  Therefore I'm wondering whether I could update the documentation to help users solve this issue.  However, your fix would work for all users without them having to add their own code, so should make exports faster for all.

I added some comments - see what you think.

File: import_export/resources.py [hunk 1]
```diff
@@ -1196,9 +1197,16 @@ def field_from_django_field(cls, field_name, django_field, readonly):
 
         FieldWidget = cls.widget_from_django_field(django_field)
         widget_kwargs = cls.widget_kwargs_for_field(field_name, django_field)
+
+        attribute = field_name
+        column_name = field_name
+        if isinstance(django_field, ForeignKey) and "__" not in column_name:
```
Comment (matthewhegarty) [inline import_export/resources.py:1203]: A comment referencing the issue could be helpful for future maintainers to understand why the code was added.

Comment (PetrDlouhy) [reply import_export/resources.py:1204]: I have added the comment.

**Comment by PetrDlouhy** [comment]
@matthewhegarty The `select_related()` makes the query slower (usually not too much) and in the case of getting object IDs it is not necessary. I would be a bit careful about telling users to add `select_related()` as temporary fix, because they will forget about it after proper fix.

Of course it is good idea to add `select_related()` in all other cases of getting fields from foreign key. The IDs are the only case where it is not necessary and also unexpected.

File: tests/core/tests/test_resources/test_resources.py [hunk 2]
```diff
@@ -289,6 +289,24 @@ def test_export(self):
             dataset = self.resource.export(queryset=Book.objects.all())
             self.assertEqual(len(dataset), 1)
 
+    @ignore_widget_deprecation_warning
+    def test_export_with_foreign_keys(self):
+        """
+        Test that export() containing foreign keys doesn't generate
+        extra query fore every row.
```
Comment (matthewhegarty) [inline tests/core/tests/test_resources/test_resources.py:296]: typo on this line ('fore')

**Review by matthewhegarty — CHANGES_REQUESTED** [review]
Thanks for clarifying.
I am happy to merge this.  Would you mind updating the changelog?

File: tests/core/tests/test_resources/test_resources.py (hunk 2 above)
Comment (PetrDlouhy) [reply tests/core/tests/test_resources/test_resources.py]: Fixed

**Comment by PetrDlouhy** [comment]
I have added new entry to changelog.

**Review by matthewhegarty — APPROVED** [review]
Thanks! 💯

File: import_export/resources.py [hunk 3]
```diff
@@ -1196,9 +1197,17 @@ def field_from_django_field(cls, field_name, django_field, readonly):
 
         FieldWidget = cls.widget_from_django_field(django_field)
         widget_kwargs = cls.widget_kwargs_for_field(field_name, django_field)
+
+        attribute = field_name
+        column_name = field_name
+        # To solve #974
+        if isinstance(django_field, ForeignKey) and "__" not in column_name:
+            attribute += "_id"
+            widget_kwargs["key_is_id"] = True
```
Comment (19greg96) [inline import_export/resources.py:1206]: Correct me if I am wrong, please, but to me, it feels like this belongs more in `widget_kwargs_for_field` than here.

**Pull request merged**